File: Kotlin/main/appendix/ThreadPools.kt
package appendix

import kotlinx.coroutines.*
import mocks.handleConnection
import mocks.Socket
import java.util.concurrent.ExecutorService

object ThreadPools {
  fun demo1(exec: ExecutorService, socket: Socket) = exec.execute {
    handleConnection(socket)
  }

  fun demo2(exec: ExecutorService, socket: Socket) = runBlocking {
    val dispatcher = exec.asCoroutineDispatcher()
    withContext(dispatcher) {
      // can use context implicitly to run coroutines
      handleConnection(socket)
    }
  }

  fun demo3(socket: Socket) {
    Dispatchers.Default.asExecutor().execute {
      handleConnection(socket)
    }
  }
}

File: Scala/main/appendix/ThreadPools.scala
package appendix

import mocks.Appendix.{ Socket, handleConnection }

import java.util.concurrent.ExecutorService
import scala.concurrent.{ ExecutionContext, Future }

object ThreadPools:
   def demo1(exec: ExecutorService, socket: Socket): Unit =
      exec.execute(() => handleConnection(socket))

   def demo2(exec: ExecutorService, socket: Socket): Unit =
      given ExecutionContext = ExecutionContext.fromExecutor(exec)

      Future {
         handleConnection(socket) // can use context implicitly to create futures
      }

   def demo3(socket: Socket): Unit =
      ExecutionContext.global.execute(() => handleConnection(socket))
end ThreadPools
